### DISCUSSION TITLE
constrain total time between node and its neighbor

### DISCUSSION THREAD

**Discussion by jmintz4** [google/or-tools#3550] [Routing (and legacy CP) questions] [answered]
What is the proper way to add a constraint between a node and its downstream (or upstream) neighbor?

e.g., the total time between node 8 and its next node must be >= 20: 

`routing.solver().Add((time_dimension.CumulVar(routing.NextVar(manager.NodeToIndex(8))) - time_dimension.CumulVar(manager.NodeToIndex(8))) >= 20)`

The above returns 
TypeError: in method 'RoutingDimension_CumulVar', argument 2 of type 'int64_t'

**Comment by Mizux** [comment, marked as answer]
`routing.NextVar(manager.NodeToIndex(8))` return an IntVar since the solver can't know the value until a solution is found while dimension::CumulVar take an index as argument thus the error.

Suppose you have a boolean expression `cond[Y] = routing.NextVar(K) == Y`
then for any value N but one you'll have `cond[i] == 0`.

so `cond[i] * (time_dimension.CumulVar(i) - time_dimension.CumulVar(K))`  is 0 for any indices except the node after K.
so `sum([cond[i] * (dim.CumulVar(i) - dim.CumulVar(K)) for i in N]) >= 20`

**Reply by jmintz4** [reply]
```python
N = len(data['time_matrix'])
K = 11
L = 20
routing.solver().Add(
  sum([
    (routing.NextVar(manager.NodeToIndex(K)) == i) * \
    (time_dimension.CumulVar(manager.NodeToIndex(i)) - time_dimension.CumulVar(manager.NodeToIndex(K))) \
    for i in range(N)]) \
  >= L)
```

Works! Thank you!
